kernel/mman.c

```diff
@@ -617,6 +619,66 @@ static int _munmap(myst_mman_t* mman, void* addr, size_t length)
     return ret;
 }
 
+static int _mprotect_delayed_zero_fill(
+    myst_mman_t* mman,
+    void* addr,
+    size_t length,
+    int prot)
+{
+    uint32_t i;
+    uint32_t start_page_index = ((uintptr_t)addr - mman->start) / PAGE_SIZE;
+    uint32_t npages = length / PAGE_SIZE;
+    bool w_access_requested = false;
+
+    if (prot != MYST_PROT_NONE)
+    {
+        for (i = 0; i < npages; i++)
+        {
+            if (mman->prot_vector[start_page_index + i] &
+                MYST_PENDING_ZEROING_FLAG)
+            {
+                if (!w_access_requested)
+                {
+                    if (myst_tcall_mprotect(
+                            addr, length, (prot | MYST_PROT_WRITE)))
+                    {
+                        _mman_set_err(mman, "mprotect tcall failed");
+                        return -EINVAL;
+                    }
+                    w_access_requested = true;
+                }
+                memset((void*)((uintptr_t)addr + i * PAGE_SIZE), 0, PAGE_SIZE);
+            }
+        }
+
+        if ((!(prot & MYST_PROT_WRITE) && w_access_requested) ||
+            (!w_access_requested))
+        {
+            if (myst_tcall_mprotect(addr, length, prot))
+            {
+                _mman_set_err(mman, "mprotect tcall failed");
+                return -EINVAL;
+            }
+        }
+        memset(&mman->prot_vector[start_page_index], prot, npages);
+    }
+    else
+    {
+        if (myst_tcall_mprotect(addr, length, prot))
+        {
+            _mman_set_err(mman, "mprotect tcall failed");
+            return -EINVAL;
+        }
+        /* Don't clear any pending zero-fill flag */
+        for (i = 0; i < npages; i++)
+        {
+            mman->prot_vector[start_page_index + i] &=
```

Comment:
Are we delaying the "delayed 0-fill" to a future `mprotect`? It takes me a while to figure out the intention. Please add comment.

Reply:
Is the loop equivalent to `memset(&mman->prot_vector[start_page_index], prot & MYST_PENDING_ZEROING_FLAG, npages);`? or `_MMAN_SET_PAGES_PROT(mman, addr, length, prot & MYST_PENDING_ZEROING_FLAG)`?

Reply:
Added comment. The loop preserve elements of MYST_PENDING_ZEROING_FLAG and set other elements to MYST_PROT_NONE (value 0). It can't be replaced by one memset().